packages/components/src/presentation-components/tree/controlled/UsePresentationTreeState.ts

```diff
@@ -208,22 +216,17 @@ interface PresentationTreeFilteringProps {
 
 function useTreeState(props: UseTreeStateProps) {
   const [state, setState] = useState<TreeState>();
-  const onNodeLoadedRef = useRef(props.onNodeLoaded);
-  useEffect(() => {
-    onNodeLoadedRef.current = props.onNodeLoaded;
-  }, [props.onNodeLoaded]);
-
-  const prevStateRef = useRef(state);
-  useEffect(() => {
-    prevStateRef.current = state;
-  }, [state]);
+  const onNodeLoadedRef = useLatest(props.onNodeLoaded);
+  const onHierarchyLimitExceededRef = useLatest(props.onHierarchyLimitExceeded);
+  const prevStateRef = useLatest(state);
 
   useEffect(() => {
     const { treeModel, ...providerProps } = props.treeStateProps;
     const modelSource = new TreeModelSource(new MutableTreeModel(treeModel));
-    const dataProvider = new PresentationTreeDataProvider(providerProps);
+    const dataProvider = new PresentationTreeDataProvider({ ...providerProps, onHierarchyLimitExceeded: () => onHierarchyLimitExceededRef.current?.() });
     const pagedLoader = new PagedTreeNodeLoader(dataProvider, modelSource, providerProps.pagingSize);
-    const nodeLoader = new ReportingTreeNodeLoader(pagedLoader, () => onNodeLoadedRef.current);
+    const nodeLoader = new ReportingTreeNodeLoader(pagedLoader, (nodeLoadedProps) => onNodeLoadedRef.current?.(nodeLoadedProps));
+    const prevState = prevStateRef.current;
```

Comment:
Why is this needed?

Reply:
ESLint complains about the cleanup function: `The ref value 'prevStateRef.current' will likely have changed by the time this effect cleanup function runs. If this ref points to a node rendered by React, copy 'prevStateRef.current' to a variable inside the effect, and use that variable in the cleanup function.`.

Reply:
Can you add eslint ignore for this and revert to how it was before? It should use changed (latest) value when cleanup runs.

Reply:
Done.